### ISSUE TITLE
Upgrade nodejs SDK to use ts-node ^10.2

### ISSUE THREAD

**Issue by delitescere** [pulumi/pulumi#7758]
The version of `ts-node` used by pulumi (7.0) is well out of date (3 years) and doesn't support certain capabilities (like importing custom local TypeScript modules for YAML files).

Forcing pulumi to use ts-node 10.2.0 (using `npm-force-resolutions`) has not caused any apparent issues.

## Steps to reproduce

1. Create a simple `.d.ts` module file for a YAML format
2. Import that into pulumi `index.ts`
3. `pulumi preview`

`example.yaml.d.ts`:
```typescript
declare module 'example.yaml' {
  const data: {
     [key: string]: any
  };

  export default data;
}
```

`index.ts`:
```typescript
import ExampleYaml from './example.yaml';

const yamlData: typeof ExampleYaml = {};
```

During `pulumi preview`:

Expected: No error, the module is imported

Actual:
```shell
    TSError: ⨯ Unable to compile TypeScript:
    index.ts: error TS2307: Cannot find module 'example.yaml'.
```

Forcing pulumi to use ts-node 10.2 works around this issue (but I'd rather not!)

**Comment by justinvp** [comment]
This is a duplicate of #4876. Note that we had upgraded to `ts-node@^8.0.0` back when that was the latest version, but it introduced a significant performance regression, so we reverted back to `ts-node@^7.0.0`. It's definitely worth revisiting this to see if perf has improved in the most recent version of `ts-node`.

**Comment by ffMathy** [comment]
@justinvp if it's worth looking into if the performance has been fixed in more recent versions of ts-node, why on earth was this PR closed? Isn't that exactly what it's doing?

**Comment by justinvp** [comment]
@ffMathy, which PR? This is an issue that was closed as a duplicate of #4876, which is tracking this.

**Comment by ffMathy** [comment]
> @ffMathy, which PR? This is an issue that was closed as a duplicate of #4876, which is tracking this.

Oh my - I am so sorry. I jumped around between PRs and issues, and thought this was a PR that solved something.

I guess I should check twice next time before I make an embarassing comment ❤️ 🤦
